File: src/main/kotlin/frc/team2186/robot/common/Motor.java
package frc.team2186.robot.common;

import edu.wpi.first.wpilibj.PowerDistributionPanel;
import edu.wpi.first.wpilibj.SpeedController;
import edu.wpi.first.wpilibj.VictorSP;
import frc.team2186.robot.Config;
import frc.team2186.robot.Robot;

public class Motor implements SpeedController {
    private VictorSP motor;
    private PowerDistributionPanel pdp;

    private int pdpChannel = 0;
    private double peakCurrent = 0.0;

    private double setpoint = 0.0;
    public Motor(Config.Grabber.Motor cfg) {
        this.motor = new VictorSP(cfg.getId());
        this.pdp = Robot.Companion.getPdp();
        this.pdpChannel = cfg.getPdpChannel();
        this.peakCurrent = cfg.getPeakCurrent();
    }

    @Override
    public void set(double speed) {
        setpoint = speed;
        motor.set(setpoint);
    }

    @Override
    public double get() {
        return setpoint;
    }

    @Override
    public void setInverted(boolean isInverted) {
        motor.setInverted(isInverted);
    }

    @Override
    public boolean getInverted() {
        return motor.getInverted();
    }

    @Override
    public void disable() {
        motor.disable();
    }

    @Override
    public void stopMotor() {
        motor.stopMotor();
    }

    @Override
    public void pidWrite(double output) {
        setpoint = output;
        motor.pidWrite(output);
    }

    public double getCurrent() {
        return pdp.getCurrent(pdpChannel);
    }

    public boolean isStalled() {
        return getCurrent() >= peakCurrent;
    }
}

File: src/main/kotlin/frc/team2186/robot/common/Motor.kt
package frc.team2186.robot.common

import edu.wpi.first.wpilibj.PowerDistributionPanel
import edu.wpi.first.wpilibj.SpeedController
import edu.wpi.first.wpilibj.VictorSP
import frc.team2186.robot.Config
import frc.team2186.robot.Robot

class Motor(cfg: Config.Grabber.Motor) : SpeedController {
    private val motor: VictorSP = VictorSP(cfg.id)
    private val pdp: PowerDistributionPanel = Robot.pdp

    private val pdpChannel = cfg.pdpChannel
    private val peakCurrent = cfg.peakCurrent

    private var setpoint = 0.0

    val current: Double
        get() = pdp.getCurrent(pdpChannel)

    val isStalled: Boolean
        get() = current >= peakCurrent

    override fun set(speed: Double) {
        setpoint = speed
        motor.set(setpoint)
    }

    override fun get(): Double {
        return setpoint
    }

    override fun setInverted(isInverted: Boolean) {
        motor.inverted = isInverted
    }

    override fun getInverted(): Boolean {
        return motor.inverted
    }

    override fun disable() {
        motor.disable()
    }

    override fun stopMotor() {
        motor.stopMotor()
    }

    override fun pidWrite(output: Double) {
        setpoint = output
        motor.pidWrite(output)
    }
}
